### PR TITLE
8323186: A faster algorithm for MutablebigInteger.divWord(long, int)

### PR DESCRIPTION
**Pull request by fabioromano1**
The method `MutableBigInteger.divWord(long, int)` can use the algorithm of Hacker's Delight (2nd ed), section 9.3, the same used in `Long.divideUnsigned(long, long)` and `Long.remainderUnsigned(long, long)`, to get the computation faster.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8323186](https://bugs.openjdk.org/browse/JDK-8323186): A faster algorithm for MutablebigInteger.divWord(long, int) (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17291/head:pull/17291` \
`$ git checkout pull/17291`

Update a local copy of the PR: \
`$ git checkout pull/17291` \
`$ git pull https://git.openjdk.org/jdk.git pull/17291/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17291`

View PR using the GUI difftool: \
`$ git pr show -t 17291`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17291.diff">https://git.openjdk.org/jdk/pull/17291.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17291#issuecomment-1881186023)

### CODE REVIEW

**Comment by bplb** [comment]
Do you have any benchmark results demonstrating the increased throughput?

**Comment by fabioromano1** [comment]
> Do you have any benchmark results demonstrating the increased throughput?

I just uploaded a class for random tests. But during the execution, I've found a non banal problem concerning the old implementation's running time.
The problem is this: the running time of the old implementation depends strongly by the difference between the initial approximation and the exact value of the quotient. So, I tried to construct of a possible worst case, attempting to maximizing this distance. To do so, I maximized the dividend `n = -2` (treated as unsigned) and minimized the divisor `d = 3`. The result is this:
The computed initial approximation is `q0 = (n >>> 1) / (d >>> 1)`, so `q0 == 9223372036854775807L`, while the exact quotient is `q = 6148914691236517204L`. This implies that the number of iterations is  `|q - q0| == 3074457345618258603L` in the worst case. This means that the running time of the old implementation can be EXPONENTIAL in the bit length of |q - q0|.
Experimentally, you can notice this by the fact that if you try to do an increasingly number of divisions, the running time quickly becomes unreasonably long, this happen just over the `2^18 == 262144` divisions.

Anyway, these are my results of the tests' running.

New algorithm:

Time to do 1 divisions: PT0.002411989S
Average time for division: PT0.002411989S

Time to do 2 divisions: PT0.000003997S
Average time for division: PT0.000001998S

Time to do 4 divisions: PT0.000006194S
Average time for division: PT0.000001548S

Time to do 8 divisions: PT0.000013294S
Average time for division: PT0.000001661S

Time to do 16 divisions: PT0.000021008S
Average time for division: PT0.000001313S

Time to do 32 divisions: PT0.000042878S
Average time for division: PT0.000001339S

Time to do 64 divisions: PT0.000131778S
Average time for division: PT0.000002059S

Time to do 128 divisions: PT0.000163048S
Average time for division: PT0.000001273S

Time to do 256 divisions: PT0.000383142S
Average time for division: PT0.000001496S

Time to do 512 divisions: PT0.000319871S
Average time for division: PT0.000000624S

Time to do 1024 divisions: PT0.000413291S
Average time for division: PT0.000000403S

Time to do 2048 divisions: PT0.000591616S
Average time for division: PT0.000000288S

Time to do 4096 divisions: PT0.00135185S
Average time for division: PT0.00000033S

Time to do 8192 divisions: PT0.013735457S
Average time for division: PT0.000001676S

Time to do 16384 divisions: PT0.008213712S
Average time for division: PT0.000000501S

Time to do 32768 divisions: PT0.020442676S
Average time for division: PT0.000000623S

Time to do 65536 divisions: PT0.011344406S
Average time for division: PT0.000000173S

Time to do 131072 divisions: PT0.018728826S
Average time for division: PT0.000000142S

Time to do 262144 divisions: PT0.043171006S
Average time for division: PT0.000000164S

Time to do 524288 divisions: PT0.073636145S
Average time for division: PT0.00000014S

Time to do 1048576 divisions: PT0.120018251S
Average time for division: PT0.000000114S

Time to do 2097152 divisions: PT0.157581276S
Average time for division: PT0.000000075S

Time to do 4194304 divisions: PT0.348321185S
Average time for division: PT0.000000083S

Time to do 8388608 divisions: PT0.642987771S
Average time for division: PT0.000000076S

Time to do 16777216 divisions: PT1.104336979S
Average time for division: PT0.000000065S

Time to do 33554432 divisions: PT2.229299703S
Average time for division: PT0.000000066S

Time to do 67108864 divisions: PT4.386180114S
Average time for division: PT0.000000065S

Time to do 134217728 divisions: PT8.833464454S
Average time for division: PT0.000000065S

Time to do 268435456 divisions: PT17.504803158S
Average time for division: PT0.000000065S

Old algorithm:
Time to do 1 divisions: PT0.00590102S
Average time for division: PT0.00590102S

Time to do 2 divisions: PT0.000002469S
Average time for division: PT0.000001234S

Time to do 4 divisions: PT0.000014707S
Average time for division: PT0.000003676S

Time to do 8 divisions: PT0.00000855S
Average time for division: PT0.000001068S

Time to do 16 divisions: PT0.000027515S
Average time for division: PT0.000001719S

Time to do 32 divisions: PT0.000073262S
Average time for division: PT0.000002289S

Time to do 64 divisions: PT0.000340843S
Average time for division: PT0.000005325S

Time to do 128 divisions: PT0.00016714S
Average time for division: PT0.000001305S

Time to do 256 divisions: PT0.001197161S
Average time for division: PT0.000004676S

Time to do 512 divisions: PT0.000273558S
Average time for division: PT0.000000534S

Time to do 1024 divisions: PT0.000617932S
Average time for division: PT0.000000603S

Time to do 2048 divisions: PT0.000892402S
Average time for division: PT0.000000435S

Time to do 4096 divisions: PT0.001367862S
Average time for division: PT0.000000333S

Time to do 8192 divisions: PT1.319736346S
Average time for division: PT0.0001611S

Time to do 16384 divisions: PT8.109278716S
Average time for division: PT0.000494951S

Time to do 32768 divisions: PT0.038191561S
Average time for division: PT0.000001165S

Time to do 65536 divisions: PT3.249995508S
Average time for division: PT0.00004959S

Time to do 131072 divisions: PT0.203684211S
Average time for division: PT0.000001553S

Time to do 262144 divisions: PT9.767934912S
Average time for division: PT0.000037261S

**Comment by fabioromano1** [comment]
@rgiulietti The correctness of the algorithm simply follows by the fact that it is just a particularization of `Long.divideUnsigned(long, long)` and `Long.remainderUnsigned(long, long)`.

**Comment by rgiulietti** [comment]
Sure. But the purpose of a test is not much to check the _current_ code but to protect against erroneous changes in the future. Without a test, a later modification might introduce an undetected regression.

**Comment by fabioromano1** [comment]
@rgiulietti The method `MutablebigInteger.divWord(long, int)` is used for division of BigIntegers, and division of BigIntegers already has several tests. Anyay, in case we can use the same tests which already exist for  `Long.divideUnsigned(long, long)` and `Long.remainderUnsigned(long, long)`.

**Comment by fabioromano1** [comment]
@rgiulietti By the way, the tests for correctness of `Long.divideUnsigned(long, long)` and `Long.remainderUnsigned(long, long)` simply check the consistency between their results and those of `BigInteger.divide(BigInteger)`, so those tests are implicitly tests also for `MutablebigInteger.divWord(long, int)`.

**Comment by rgiulietti** [comment]
What about splitting the result of `divWord()` into its components `q` and `r` and check that `0 <= r < d` and `n == q * d + r`? Mathematically this should be sufficient. Perhaps you might need to take care about unsigned vs signed arithmetic.

**Comment by fabioromano1** [comment]
@rgiulietti In this case, the unsigned arithmetic is not a problem, since `d & LONG_MASK > 0` and multiplication and addition semantics is equivalent in signed/unsigned. Perhaps, it would be better to call directly `Long.divideUnsigned(long, long)` and `Long.remainderUnsigned(long, long)` to possibly use directly assembly code, since they are annotated as IntrinsicCandidate.

**Comment by rgiulietti** [comment]
AFAIK, IntrinsicCandidate methods are only relevant in JIT compiled code. A test that checks correctness might not reach the compilation stage, and execute only in the bytecode interpreter.

But IMO using the result of `divWord()` as suggested is simple enough and avoids hidden dependencies.

**Comment by fabioromano1** [comment]
> AFAIK, IntrinsicCandidate methods are only relevant in JIT compiled code. A test that checks correctness might not reach the compilation stage, and execute only in the bytecode interpreter.
> 
> But IMO using the result of `divWord()` as suggested is simple enough and avoids hidden dependencies.

@rgiulietti Perhaps I was not clear enough. I'm not intended to use `Long.divideUnsigned(long, long)` and `Long.remainderUnsigned(long, long)` in the tests, but in the algorithm implementation, instead of using an hard-coded implementation, to use directly the assembly code, although this implies to calculate the approximations of `q` and `r` two times instead of one.

**Comment by rgiulietti** [comment]
As you note, that would probably require two divisions. I don't know if the JIT compiler can detect that the arguments are the same and emit just one division instead.
I think your code is good enough for the purpose of [Mutable]BigInteger, and better than the current one.

**Comment by fabioromano1** [comment]
@rgiulietti It seems unlikely that the JIT compiler can do that, at least to me.

File: test/jdk/java/math/BigInteger/TestDivWord.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
+import java.math.Accessor;
+import java.time.Duration;
+import java.util.Random;
+
+public class TestDivWord {
```
Comment (bplb) [inline test/jdk/java/math/BigInteger/TestDivWord.java:5]: Where is this used in actually jtreg testing?

Comment (fabioromano1) [reply test/jdk/java/math/BigInteger/TestDivWord.java:5]: @bplb It is not used in jtreg testing.

Comment (bplb) [reply test/jdk/java/math/BigInteger/TestDivWord.java:5]: So it is only for verification purposes in the context of this PR?

Comment (fabioromano1) [reply test/jdk/java/math/BigInteger/TestDivWord.java:5]: @bplb Yes, it is.

**Pull request closed without merging**